File: lrpc-java/src/com/lrpc/tcp/Service.java
package com.lrpc.tcp;

import java.io.IOException;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.ArrayList;
import com.lrpc.buf.RecvBuf;
import com.lrpc.buf.SendData;
import com.lrpc.fun.Fun;

/**
 * synchronous TCP receiving data calling registration method
 */
public class Service implements java.io.Closeable {

    private Fun srvFun;
    private ArrayList<ServerSocket> srvs;

    /**
     * @param fun call the function with Fun
     */
    public Service(Fun fun) {
        srvFun = fun;
        srvs = new ArrayList<>();
    }

    /**
     * close all ServerSockets
     */
    public void close() {
        for (ServerSocket srv : srvs) {
            try {
                srv.close();
            } catch (IOException _e) {
            }
        }
    }

    /**
     * this method is blocked for running services
     * 
     * @param port listening port
     */
    public void run(int port) throws IOException {
        ServerSocket srv = new ServerSocket(port);
        srv.setReuseAddress(true);
        srvs.add(srv);
        while (true) {
            Socket socket = srv.accept();
            new Thread(() -> {
                try {
                    byte[] buf = new byte[1024];
                    while (true) {
                        RecvBuf recv = new RecvBuf();
                        while (true) {
                            if (recv.size() != null && recv.size() == recv.len()) {
                                break;
                            }
                            int read = socket.getInputStream().read(buf);
                            if (read > 0) {
                                recv.append(buf, read);
                            } else {
                                socket.shutdownInput();
                                socket.shutdownOutput();
                                socket.close();
                                return;
                            }
                        }
                        socket.getOutputStream().write(new SendData(srvFun.invoke(recv.byteQue())).toArray());
                    }
                } catch (Exception _e) {
                    try {
                        socket.shutdownInput();
                        socket.shutdownOutput();
                        socket.close();
                    } catch (Exception __e) {
                    }
                }
            }).start();
        }
    }
}

File: lrpc-csharp/src/lrpc/tcp/Service.cs
using System;
using System.Collections.Generic;
using System.Net.Sockets;
using System.Threading;
using System.Net;
using lrpc.fun;
using lrpc.buf;

namespace lrpc.tcp
{
    /// <summary>
    /// synchronous TCP receiving data calling registration method
    /// </summary>
    public class Service : IDisposable
    {
        private Fun srvFun;
        private List<Socket> srvs;

        /// <param name="fun">call the function with Fun</param>
        public Service(Fun fun)
        {
            srvFun = fun;
            srvs = new List<Socket>();
        }

        /// <summary>
        /// close all Sockets
        /// </summary>
        public void Dispose()
        {
            foreach (Socket srv in srvs)
            {
                srv.Dispose();
            }
        }

        /// <summary>
        /// this method is blocked for running services
        /// </summary>
        /// <param name="port">listening port</param>
        public void Run(int port)
        {
            Socket srv = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            srv.ExclusiveAddressUse = false;
            srv.Bind(new IPEndPoint(IPAddress.Any, port));
            srv.Listen(int.MaxValue);
            srvs.Add(srv);
            while (true)
            {
                Socket socket = srv.Accept();
                new Thread(() =>
                {
                    try
                    {
                        byte[] buf = new byte[1024];
                        while (true)
                        {
                            RecvBuf recv = new RecvBuf();
                            while (true)
                            {
                                if (recv.Size.HasValue && recv.Size.Value == recv.Len)
                                {
                                    break;
                                }
                                int read = socket.Receive(buf);
                                if (read > 0)
                                {
                                    recv.Append(buf, read);
                                }
                                else
                                {
                                    socket.Shutdown(SocketShutdown.Both);
                                    socket.Dispose();
                                    return;
                                }
                            }
                            byte[] rst = new SendData(srvFun.Invoke(recv.ByteQue)).ToArray();
                            int write = 0;
                            while (true)
                            {
                                write += socket.Send(rst, write, rst.Length - write, SocketFlags.None);
                                if (write >= rst.Length)
                                {
                                    break;
                                }
                            }
                        }
                    }
                    catch
                    {
                        socket.Shutdown(SocketShutdown.Both);
                        socket.Dispose();
                    }
                }).Start();
            }
        }
    }
}
